diff --git a/src/PrevnextService.php b/src/PrevnextService.php
index 00f04c1..d471c56 100644
--- a/src/PrevnextService.php
+++ b/src/PrevnextService.php
@@ -64,13 +64,24 @@ class PrevnextService implements PrevnextServiceInterface {
     $query = $this->entityTypeManager->getStorage('node')->getQuery();
     $bundle = $node->bundle();
     $langcode = $node->language()->getId();
-    $nodes = $query->condition('status', NodeInterface::PUBLISHED)
+
+    $query->condition('status', NodeInterface::PUBLISHED)
       ->accessCheck(FALSE)
       ->condition('type', $bundle)
       ->condition('langcode', $langcode)
       ->addMetaData('type', $bundle)
-      ->addMetaData('langcode', $langcode)
-      ->addTag('prev_next_nodes_type')
-      ->execute();
+      ->addMetaData('langcode', $langcode);
+
+    // Add sort by weight field to query if weight module and field exists.
+    if (\Drupal::service('module_handler')->moduleExists('weight')) {
+      $fields = \Drupal::service('entity_field.manager')->getFieldDefinitions('node', $bundle);
+      foreach ($fields as $key => $field) {
+        if ($field->getType() == 'weight') {
+          $query->sort($key, 'ASC'); break;
+        }
+      }
+    }
+
+    $nodes = $query->addTag('prev_next_nodes_type')->execute();

     return array_values($nodes);
   }
